### PR TITLE
Implemented Watch mode filter by filename and by test name #1530

### PR DESCRIPTION
**Pull request by mmulet**
Hi!
This pull request implements `Watch mode filter by filename and by test name` from issue #1530

Watch this short video for a demonstration of the feature:

https://github.com/user-attachments/assets/160e3d45-0099-48ba-9a2b-050d7bf73590

### Overview

- [x] Added new tests for this feature in [test/watch-mode/basic-functionality.js](https://github.com/avajs/ava/compare/main...mmulet:ava:Watch_mode_filter?expand=1#diff-6cd81b4e1b3b06567813e753791a51e4bc7609478d01c4cb766b606a64f8a057)
- [x] Added documentation in the [docs/recipes/watch-mode.md](https://github.com/avajs/ava/compare/main...mmulet:ava:Watch_mode_filter?expand=1#diff-e759e03510c7ba375dbdc805e9348e62c3a3b3be18520c167a15c0d7e9751480)
- [x] Feature works as requested in #1530. Namely:
- [x] Press p to filter by a filename regex pattern.
- [x] Press t to filter by a test name regex pattern.
- [x] Press q to quit watch mode.
- [x] Press Enter to trigger a test run.
- [x] Reruns the filtered list of tests after applying filter (just like jest does)
- [x] Use a to rerun all tests.

<details>
  <summary>Summary of new feature</summary>

(taken from the docs)

# Filter tests while watching

You may also filter tests while watching by using the cli. For example, after running
```console
$ npx ava --watch
```
You will see a prompt like this :
```console
Type `p` and press enter to filter by a filename regex pattern
	[Current filename filter is $pattern]
Type `t` and press enter to filter by a test name regex pattern
	[Current test filter is $pattern]

[Type `a` and press enter to run *all* tests]
(Type `r` and press enter to rerun tests ||
	Type \`r\` and press enter to rerun tests that match your filters)
Type `u` and press enter to update snapshots

command > 
```
So, to run only tests numbered like
- foo23434
- foo4343
- foo93823

You can type `t` and press enter, then type `foo\d+` and press enter.
This will then run all tests that match that pattern.
Afterwards you can use the `r` command to run the matched tests again,
or `a` command to run **all** tests.

</details>

<details>
  <summary>Code Overview</summary>

In the [contributing guidelines](https://github.com/avajs/ava/blob/024de32a15aba9fac4002aa8d3c2c076955f44f4/.github/CONTRIBUTING.md), you say that you prefer shorter pull requests, but this is a bit long. So, to help speed things along, I'm going to provide an overview for each part that I added or changed.

1. The root of the feature starts in [lib/watcher.js](https://github.com/avajs/ava/compare/main...mmulet:ava:Watch_mode_filter?expand=1#diff-6e617ee9deaa8b3d2dcc52856362425d5ee4997c9d41013872cd9447588e987e)

I add a new variable called watchModeSkipTests
https://github.com/mmulet/ava/blob/78a3768fbe3360930351374f854f9cc7a46fda51/lib/watcher.js#L104

which is an instance of WatchModeSkipTests [lib/watch-mode-skip-tests.js](https://github.com/avajs/ava/compare/main...mmulet:ava:Watch_mode_filter?expand=1#diff-0459d8ffdd0edd259872b193989f2b9f4fdbe2a821a30de23f1e32e84b40d7dd), this class handles the regex for the skipping, and most of the code will be passing this object around, populating its fields, etc.

There is one particular part of this file that I want to draw attention to.

</details>

<details>
  <summary>#OnDataGenerator</summary>

Continue down the `plan` function until you come to the const _class;
This class handles the prompt for commands at watch mode. It does this by listening to stdin's data event in an asynchronous generator. 

I switched to using an asynchronous generator rather than just plain old stdin.on('data') because some new commands (namely p, and t the ones where you enter regex), also need to listen to 'data'. By this I mean that they have to listen for data while processing the command. In other words, without the generator you would need an explicit state machine which can be hard to debug and keep track of.

To see what I mean, look at listenForCommand:
```javascript
#data;
constructor() {
	this.#data = this.#onDataGenerator();
	this.#listenForCommand();
}
#onDataGenerator = async function * () {
...
}

#listenForCommand = async () => {
	for await (const data of this.#data) {
		await this.#onCommand(data);
	}
};
```
Note that the `for await` waits on #data, not on this.#onDataGenerator()! What this means is that nested inside the #onCommand function we can await more lines of input, like this:

```javascript
#onCommand = async (data) => {
  switch(data) {
     case 't':
        promptTheUserForSomething()
        const dataFromUser = await this.#data.next();
        askTheUserAnotherQuestion();
        const dataFromUserAgain = await this.#data.next();
        ...
```
Then when you are done processing your command, ie, the this.#onCommand returns, you can run the next loop (by getting the next this.#data) and everything works as it should.

```javascript
#listenForCommand = async () => {
        //This works just as you think it would,
        // each data is a new command, not
        // a prompt or something having to do with
        // inner state
	for await (const data of this.#data) {
		await this.#onCommand(data);
	}
};
```

And, that's how it works. 
It's not too complicated, but maybe a bit unusual. So, I just wanted to document my choices and communicate clearly about why I made them.

The rest of the addition to the plan function is as you would expect, it asks the user for the command and prompts for regex filters if necessary.

</details>

<details>
  <summary>Oh the places you'll go</summary>

As I mentioned before, the rest of the code is mostly passing around the watchModeSkipTests. 
- From the plan function, the watchModeSkipTests, gets added to `instructions` which is [yielded](https://github.com/mmulet/ava/blob/78a3768fbe3360930351374f854f9cc7a46fda51/lib/watcher.js#L608)
- The yield ends up at the [async for loop in start](https://github.com/mmulet/ava/blob/78a3768fbe3360930351374f854f9cc7a46fda51/lib/watcher.js#L75), which means it gets rolled up in the `runtimeOptions` variable. Which is the passed to api.run 
- From inside api.run, the watchModeSkipTestsOrUndefined branches out to 2 different places
1. First is to the [fork options](https://github.com/mmulet/ava/blob/78a3768fbe3360930351374f854f9cc7a46fda51/lib/api.js#L272), which passes it to the fork and in turn the worker. Note that the fork options is `watchModeSkipTestsData` because only the Data gets passed to the worker, not the class itself.
2. Then, it gets [emitted as part of the 'run' event](https://github.com/mmulet/ava/blob/78a3768fbe3360930351374f854f9cc7a46fda51/lib/api.js#L198). This run event will be consumed by the reporter, more on that later.

Since we are branching, let's choose a path
#### To the worker
- So, after api.run, we can let the existing code pass all the options to the worker, it's only in [worker/base.js]
(https://github.com/mmulet/ava/blob/78a3768fbe3360930351374f854f9cc7a46fda51/lib/worker/base.js#L75) that we need to pass the Data to the Runner
- Once we are in the runner, we've reached the skipTests 1 of 2 [destinations](https://github.com/mmulet/ava/blob/78a3768fbe3360930351374f854f9cc7a46fda51/lib/runner.js#L33)
It is inside the runner that we use the watchModeSkipTests to actually , well, skip tests. The logic is simple, on [every chain](https://github.com/mmulet/ava/blob/78a3768fbe3360930351374f854f9cc7a46fda51/lib/runner.js#L99) check if we should skip the test, and if we should, we skip!

#### Another Side, Another Story
Meanwhile, we have a second task, not just skip a test from running, we also want to prevent its result from being shown at all (this is a filter, the whole idea is we have too many tests to sift through).
- Let's pick up from where we left off, being emitted by emit('run'), this brings us to the [cli.js](https://github.com/mmulet/ava/blob/78a3768fbe3360930351374f854f9cc7a46fda51/lib/cli.js#L488), all the arguments for run are in the plan variable which gets sent to the reporter with the reporter.startRun(plan).
- At startPlan, we've reached our destination, [the reporter](https://github.com/mmulet/ava/blob/78a3768fbe3360930351374f854f9cc7a46fda51/lib/reporters/default.js#L142).
- Now, recall from our other destination (the runner) that we skip all tests that have been filtered out (don't match our regexes). So to prevent the output from being displayed, we look at all tests being marked [skip] and [filter out those that are being skipped due to our filters](https://github.com/mmulet/ava/blob/78a3768fbe3360930351374f854f9cc7a46fda51/lib/reporters/default.js#L255)

And there you have it, a complete tour of the changes!

</details>

<details>
  <summary>About the tests</summary>

One thing you'll notice while reviewing the tests is that I added [try catch blocks on each and every one](https://github.com/mmulet/ava/blob/78a3768fbe3360930351374f854f9cc7a46fda51/test/watch-mode/basic-functionality.js#L188) . I did this because whenever a test would fail, the entire test would just hang and end in a timeout (presumably because it throws an error, and the this.done function never gets run).

I tried this with a simple example, and it also had the same result:

```javascript

test('basic_test', withFixture('basic'), async (t, fixture) => {
	await fixture.watch({
		async 1({process}) {
			t.is(1,2); //should fail
			this.done();
		}
	});
```

Your contributing guidelines say not to include unrelated code changes in the pull request, so I'm considering this error as out of scope for now. But, now you know why the try catch blocks are there (the CLI will still report a `t.$someFunc()` failure as a failure, it just won't report exceptions as errors, `throw $someError` will be caught)

</details>

### Last but not least
I'm going to submit this pull request for the bounty on https://oss.issuehunt.io/r/avajs/ava/issues/1530, but any extra tips (you
can use [GitHub Sponsor page](https://github.com/sponsors/mmulet)) would be very welcome!

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#1530: Watch mode filter by filename and by test name](https://oss.issuehunt.io/repos/26820798/issues/1530)
---
</details>

### CODE REVIEW

**Comment by novemberborn** [comment]
Hey @mmulet, this sounds really great! I'm a bit short on time but can hopefully have a look in the next few days.

**Comment by novemberborn** [comment]
High-level question, Jest uses regular expressions to select files and tests? I ask because our CLI options use glob patterns for files, and [wildcard matching](https://www.npmjs.com/package/matcher) for individual tests. I'm wondering if it's more valuable to follow our CLI behavior, or to match Jest's interface.

And the regular expressions are matched agains filenames, rather than paths?

Also we refer to test "titles", not "names".

File: lib/watch-mode-skip-tests.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		this.fileRegexOrNull = watchModeSkipTestsData.fileRegexOrNull;
+		this.testRegexOrNull = watchModeSkipTestsData.testRegexOrNull;
+	}
+
+	shouldSkipFile(file) {
+		if (this.fileRegexOrNull === null) {
+			return false;
+		}
+
+		return !this.fileRegexOrNull.test(file);
```
Comment (novemberborn) [inline lib/watch-mode-skip-tests.js:28]: These kinds of invocations can be shortened like so:

```suggestion
		return this.#fileRegex?.test(file) !== true;
```

**Comment by novemberborn** [comment]
> One thing you'll notice while reviewing the tests is that I added [try catch blocks on each and every one](https://github.com/mmulet/ava/blob/78a3768fbe3360930351374f854f9cc7a46fda51/test/watch-mode/basic-functionality.js#L188) . I did this because whenever a test would fail, the entire test would just hang and end in a timeout (presumably because it throws an error, and the this.done function never gets run).

Oh thanks for the reminder, I think I've suffered through that in the past. I'll push up a fix.

**Comment by mmulet** [comment]
> High-level question, Jest uses regular expressions to select files and tests? I ask because our CLI options use glob patterns for files, and [wildcard matching](https://www.npmjs.com/package/matcher) for individual tests. I'm wondering if it's more valuable to follow our CLI behavior, or to match Jest's interface.


1. Confirmed yes, jest uses regex pattern. Doesn't make a difference to me whether is uses regex or glob.
Screenshot 2025-04-04 at 10 50 26 AM

> And the regular expressions are matched agains filenames, rather than paths?

2. Just tested it in jest again, and jest matches against **paths**, not just filenames. I believe this the implementation I wrote for ava also matches against the whole path.

Screenshot 2025-04-04 at 10 49 58 AM

> Also we refer to test "titles", not "names".

3. My bad.

File: lib/runner.js [hunk 2]
```diff
@@ -91,6 +96,11 @@ export default class Runner extends Emittery {
 			metadata.taskIndex = this.nextTaskIndex++;
 
 			const {args, implementation, title} = parseTestArgs(testArgs);
+			if (
+				this.shouldSkipThisTestBecauseItDoesNotMatchARegexSetAtTheWatchCLI(metadata, title.value)
+			) {
+				metadata.skipped = true;
```
[the remark this answers is not in the document]
Comment (mmulet) [reply lib/runner.js:102]: Like this?
https://github.com/user-attachments/assets/52eb50df-2db0-4ea5-a0b5-02d8def43f30

From my simple test, it doesn't appear to work at all. It looks like it reruns all tests. Or are looks deceiving? 

When I have more time, I can adjust the tests to be nondeterministic (or have some easily observable side effect) so I'll know if it's really running again or not.

File: lib/runner.js [hunk 3]
```diff
@@ -180,7 +190,7 @@ export default class Runner extends Emittery {
 		}, {
 			serial: false,
 			exclusive: false,
-			skipped: false,
+			skipped: this.skipAllTestsInThisFile,
```
[the remark this answers is not in the document]
Comment (mmulet) [reply lib/runner.js:193]: Agreed on selected, especially if we make the above change.

Skipping the entire file load does make sense. But, are there any instances that would break any existing tests? For example, what if the test file has some side effect that affects tests in another test file (obviously not a good practice, but I've seen people do unbelievable things)

File: lib/runner.js [hunk 4]
```diff
@@ -551,4 +561,12 @@ export default class Runner extends Emittery {
...
 		this.interrupted = true;
 	}
+
+	shouldSkipThisTestBecauseItDoesNotMatchARegexSetAtTheWatchCLI(metadata, testTitle) {
+		if (metadata.skipped || this.skipAllTestsInThisFile) {
+			return true;
+		}
+
+		return this.watchModeSkipTests.shouldSkipTest(testTitle);
+	}
```
[the remark this answers is not in the document]
Comment (mmulet) [reply lib/runner.js:571]: Good point.

File: test/watch-mode/fixtures/filter-files/test1.test.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+const {default: test} = await import(process.env.TEST_AVA_IMPORT_FROM); // This fixture is copied to a temporary directory, so import AVA through its configured path.
+
+test("alice", (t) => {
+  t.pass();
+});
+
+test("bob", async (t) => {
+  const bar = Promise.resolve("bar");
+  t.is(await bar, "bar");
```
[the remark this answers is not in the document]
Comment (mmulet) [reply test/watch-mode/fixtures/filter-files/test1.test.js:9]: Agreed

File: lib/watcher.js [hunk 6]
```diff
@@ -11,14 +12,49 @@ import {
...
 
 const debug = createDebug('ava:watcher');
 
 // In order to get reliable code coverage for the tests of the watcher, we need
 // to make Node.js write out interim reports in various places.
 const takeCoverageForSelfTests = process.env.TEST_AVA ? v8.takeCoverage : undefined;
 
-const END_MESSAGE = chalk.gray('Type `r` and press enter to rerun tests\nType `u` and press enter to update snapshots\n');
+const endMessage = skipTests =>
+	chalk.gray(
```
[the remark this answers is not in the document]
Comment (mmulet) [reply lib/watcher.js:24]: Yeah, makes sense.

File: lib/watcher.js [hunk 7]
```diff
@@ -37,15 +73,36 @@ export function available(projectDir) {
...
+	api,
+	filter,
+	globs,
+	projectDir,
+	providers,
+	stdin,
+	abortSignal,
+	reporter,
+}) {
+	const watchModeSkipTests = new WatchModeSkipTests();
```
[the remark this answers is not in the document]
Comment (mmulet) [reply lib/watcher.js:104]: Sounds fine to me.

File: lib/watcher.js [hunk 8]
```diff
@@ -149,18 +207,159 @@ async function * plan({api, filter, globs, projectDir, providers, stdin, abortSi
...
+	 * for lines of data from stdin.
+	 *
+	 * We store this generator in the field
+	 * #data which we listen to in both
+	 * the #listenForCommand function,
+	 * and the promptForFilter. By keeping
+	 * one generator instance for both
+	 * we can have nested listeners.
+	 */
+	const _class = new (class {
```
[the remark this answers is not in the document]
Comment (mmulet) [reply lib/watcher.js:229]: I don't think it's a need, we can just declare everything as variables in the plan function scope and it should work just fine.

File: lib/watcher.js [hunk 9]
```diff
@@ -149,18 +207,159 @@ async function * plan({api, filter, globs, projectDir, providers, stdin, abortSi
...
+		#data;
+
+		constructor() {
+			this.#data = this.#onDataGenerator();
+			this.#listenForCommand();
 		}
-	});
+
+		#onDataGenerator = async function * () {
```
[the remark this answers is not in the document]
Comment (mmulet) [reply lib/watcher.js:237]: I haven't used that lib before, but I took a look at the docs. Ideally we could something like this:

```javascript
   #onDataGenerator = async function * () {
        for await (const line of stdin.pipe(split2())) {
            yield line.trim();
        }
    };
```

[the remark this answers is not in the document]
Comment (mmulet) [reply lib/watcher.js:237]: I just tested it, and this works great!

**Comment by mmulet** [comment]
I believe I responded to all the suggestions. I'm out of time for now, but I should be able to work again on this Sunday night or Monday and implement these changes.

In the meantime, feel free to make any changes. Based on the suggestions you've made, we really seem to be on the same page here. Plus, I'm not the type to be possessive about code, so rewrite the whole thing if you want to.

File: lib/watch-mode-skip-tests.js (hunk 1 above)
Comment (mmulet) [reply lib/watch-mode-skip-tests.js:28]: I tested this and it doesn't work.


Counterexample:
```javascript
#fileRegex = null;

this.#fileRegex?.test(file) !== true;
---------------------------
this.#fileRegex?.test(file) // is undefined because #fileRegex is null
undefined !== true // Is true because undefined does not equal to true
// But the correct value is false!
```
This is incorrect because when the the filter is null we should *not* skip anything, the correct result is false.


A shorter version that does work is:
```javascript
!(this.#fileRegex?.test(file) ?? true);
```

File: lib/runner.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmulet) [reply lib/runner.js:102]: By using selected instead of skipped, we can take out all the modifications to the reporter, this will really simplify the code. So, yes I like `selected` much better than skipped.

**Comment by mmulet** [comment]
# Summary of changes

This should be in the same order as the suggestions given above.

- [x] Test "titles" not test names. In watch-mode-skip-tests
- [x] Test "titles" not test names. In cli 
- [x] metadata.selected instead of skipped
- [x] adjust the `basic-functionality` tests to reflect selected vs skipped
- [x] initialize to selected: false in runner.js
- [x] Skip loading entire file if we already know if we are skipping all the tests
- [x] Remove `shouldSkipThisTestBecauseItDoesNotMatchARegexSetAtTheWatchCLI` wrapper
- [x] Change tests fixture to just use `t.pass()` or `t.fail()` instead of `t.is("bar", "bar")`
- [x] Instead of `endMessage(...) : string`, use `endMessage(reporter,...) : void`
- [x] Change `watchModeSkipTests` to `InteractiveFilter`
- [x] Change `onDataGenerator` to use `split2`
- [x] Remove anonymous class, and replace with more variables
- [x] Change `watchModeSkiptTestsOrUndefined` to `InteractiveFilter` (no undefined)
- [x] Explicit api to convert `WatchModeSkipTests` to data (with getData)
- [x] Made the `InteractiveFilter` regexes private.
- [x] Remove `replace(file|test)Regex` and replace with setter
    Note, I actually didn't remove the replaceRegex function, because I added a new return value to it (filterChanged) that let's us know if the regex had any changes. This way, we don't have to rerun the tests if the regex didn't change.
- [x] `fileRegexOrNull` => `fileRegex` (no null), same for titleRegex
- [x] shorter invocation for `this.#fileRegex?.test(file) !== true;` replaced with !(this.#filepathRegex?.test(file) ?? true)

File: lib/runner.js [hunk 10]
```diff
@@ -97,13 +95,11 @@ export default class Runner extends Emittery {
 
 			const {args, implementation, title} = parseTestArgs(testArgs);
 			if (
-				this.shouldSkipThisTestBecauseItDoesNotMatchARegexSetAtTheWatchCLI(metadata, title.value)
+				this.interactiveFilter.shouldSelectTest(title.value)
```
Comment (mmulet) [inline lib/runner.js:98]: Moved to selected rather than skipped. Most of the code checks for `this.checkSelectedByLineNumbers` whenever it also checks metadata.selected. Like this:

`if(this.checkSelectedByLineNumbers !metadata.selected)`

I removed the this.checkSelectedByLineNumbers. All tests still pass, and it works fine. So, I am assuming that check is not needed.

File: lib/api.js [hunk 11]
```diff
@@ -162,6 +162,8 @@ export default class Api extends Emittery {
 			setupOrGlobError = error;
 		}
 
+		selectedFiles = selectedFiles.filter(file => runtimeOptions.interactiveFilter?.canSelectTestsInThisFile(file) ?? true);
```
Comment (mmulet) [inline lib/api.js:165]: This seemed like the best place to skip loading entire file if it was filtered out by the interactive filter, but would somewhere else be better?

File: lib/runner.js [hunk 12]
```diff
@@ -411,7 +418,7 @@ export default class Runner extends Emittery {
 				continue;
 			}
 
-			if (this.checkSelectedByLineNumbers && !task.metadata.selected) {
+			if (!task.metadata.selected) {
```
Comment (mmulet) [inline lib/runner.js:421]: When I moved to metadata.selected rather than metadata.skipped, I had to delete the checks for `this.checkSelectedByLineNumbers` whenever it also checks metadata.selected.

All tests still pass, and it works fine. So, I am assuming that check is not needed.

File: lib/runner.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (novemberborn) [reply lib/runner.js]: > Skipping the entire file load does make sense. But, are there any instances that would break any existing tests? For example, what if the test file has some side effect that affects tests in another test file (obviously not a good practice, but I've seen people do unbelievable things)

I don't think it's AVA's responsibility to support such side-effects. Equally, people may do setup in the test file directly and then teardown in an after-hook and they would be surprised when their after-hook doesn't run.

File: lib/api.js [hunk 13]
```diff
@@ -206,6 +206,7 @@ export default class Api extends Emittery {
 				runOnlyExclusive: runtimeOptions.runOnlyExclusive === true,
 				firstRun: runtimeOptions.firstRun ?? true,
 				status: runStatus,
+				watchModeSkipTestsOrUndefined: runtimeOptions.watchModeSkipTestsOrUndefined,
```
[the remark this answers is not in the document]
Comment (novemberborn) [reply lib/api.js]: > As a first time contributor, Typescript would have made contributing 100% easier, half my startup time was figuring out what type everything was and tracing the call stack to see where everything goes. With plain javascript, vscodes's "Go To Definition" was very very unreliable. So instead I had to do a lot of manually searching.

Yea, most parts of AVA are starting to show their age.

**Comment by novemberborn** [comment]
@mmulet still looking at this; there's some edge cases that go pretty deep and I wanted to see how it'd work with globs and matches, too. Getting close to having that resolved.

Edge cases:

* Not running tests while prompting for a filter
* Not clearing the command prompt line when running tests due to changes, breaking the separator
* Correctly resetting previous failure counts when filtering over all possible test files (i.e. not from a file change)
* The interplay between the command line glob filter and the interactive file filter (regex or glob)
* The interplay between the interactive filter (regex or `--match` equivalent), `--match` and `.only` test modifiers (struggling with this at the moment)

I'll push up some changes when I can, for your feedback.

Thanks for pushing this forward!

**Comment by mmulet** [comment]
Sorry for the delay, it might be a week or two until I get to this.

**Comment by mmulet** [comment]
This is going to sound anticlimactic after 3 weeks, but "Looks great!".

- I like the glob patterns better than regex, it's more consistent
- I also like that we got rid of the `InteractiveFilter` class and moved everything to `selected`. It's much cleaner.

## A few fixes
- I updated the docs to reflect the new glob patterns
- I found that if you do not have any patterns, but type `m` or `g` to enter a glob, then you leave the prompt blank, it will not return to the original prompt. The cause is `oldGlobPattern === interactiveGlobPattern` in `watcher.js`. I fixed it by always calling `signalChanged()` (moving it out of the else), but I'm not sure if this will have unintended consequences.

**Comment by novemberborn** [comment]
> This is going to sound anticlimactic after 3 weeks, but "Looks great!".

Haha, no worries, glad to hear!



> * I found that if you do not have any patterns, but type `m` or `g` to enter a glob, then you leave the prompt blank, it will not return to the original prompt.

Yes I did that on purpose actually, as a way to reset the pattern. With that insight, do you think it's a bug or a feature?

**Comment by mmulet** [comment]
> Yes I did that on purpose actually, as a way to reset the pattern. With that insight, do you think it's a bug or a feature?

It does work to clear the pattern (and that's good), but I ran into the problem when I don't yet have any patterns (or I wish to keep the pattern the same).


Let's say I type `m` and enter the prompt for a new test filter, but I change my mind (or I meant to do file filter) and want to cancel and go back. It's not immediately clear to me how to do that. My first thought is to leave the prompt blank (because the tips are telling me to leave it blank to clear the pattern) and press return, but when I do that it just enters a line and I can't go back.


For this video I reverted the changes I just made to show what I just described:

https://github.com/user-attachments/assets/5a96a12b-46e1-40e7-ae99-0a4f68ac6a8d

I have no way to escape the filter prompt without entering a filter.

**Comment by novemberborn** [comment]
> I have no way to escape the filter prompt without entering a filter.

OK that's definitely a bug.



> Let's say I type `m` and enter the prompt for a new test filter, but I change my mind (or I meant to do file filter) and want to cancel and go back. It's not immediately clear to me how to do that.

I don't think we can do a cancel, other than by having you type a magic input. The TTY isn't in raw mode (which would let us recognize say an ESC press). That's because raw mode breaks Control+C interrupts.

**Comment by mmulet** [comment]
> Oh, `readline`, yes that should be entirely appropriate for this use case! Can't believe I forgot about that.
> 
> Looks like Ctrl+D can act as a "I have no input, don't erase my pattern" signal?

That would work but unfortunately `Ctrl+D` [doesn't work on Windows](https://nodejs.org/api/readline.html?utm_source=chatgpt.com#tty-keybindings). 

Right now I added support for using `Ctrl+C`, such that `Ctrl+C` works as normal at the top of the prompt, but let's you use it as `I have no input, don't erase my pattern` while it is asking for a glob or match pattern.

**Comment by mmulet** [comment]
I think it's ready to ship!

**Comment by mmulet** [comment]
A little issue rewarding the [issue on issuehunt](https://oss.issuehunt.io/r/avajs/ava/issues/1530), it says `closed` not `solved`, and hasn't been rewarded yet. 

I don't know if the issue is on issuehunt's end, or if there is something you can do to `reward` it. In case you are unfamiliar with issuehunt, I know @sindresorhus knows how it works, because I just solved another issue about a month ago and it got rewarded. [IssueHunt Link](https://oss.issuehunt.io/r/sindresorhus/gifski-app/issues/64)   , [GitHhub Link](https://github.com/sindresorhus/Gifski/issues/219)
